IsSuperMajority() softfork for BIPs 68,112 and 113

btcdrak opened the pull request:
## General

This PR softforks 3 lock-time related BIPs which are all implemented in `master` as mempool-only logic at the moment.
1. [BIP68](https://github.com/bitcoin/bips/blob/master/bip-0068.mediawiki) sequence locks for relative locktime;
2. [BIP112](https://github.com/bitcoin/bips/blob/master/bip-01128.mediawiki) CHECKSEQUENCEVERIFY;
3. [BIP113](https://github.com/bitcoin/bips/blob/master/bip-0113.mediawiki) Media Time Past.
### Dependencies

Why 3 BIPs at once?
- BIP68 (sequence locks) independently enforces the same semantics BIP113 (MTP)
- BIP112 (CSV) relies on BIP68
### Relay policy for BIP113

BIP113 mempool-only policy was deployed with Bitcoin Core 0.11.2 at the same time as the BIP65 CLTV softfork so the policy is in wide use (at  least 42% of nodes) as well as all the miners who upgraded to 0.11.2.
### Relay policy for v2 transactions

BIP68/112 rely on v2 transactions. Currently only v1 transactions are relayed, so it is necessary to change the relay policy to allow v2. This will be done at the same time as softfork deployment and will have the net effect that once the softfork enforces, we can be pretty sure miners will mine v2 transactions.

At a later date once enough nodes upgrade and we're sure v2 transaction will be relayed efficiently, we can bump the default transaction version in core see #7562.
### Why ISM()?

It seems unlikely that BIP9 will be both developed and well tested enough to meet the roadmap schedule of deployment even with reasonable slippage taken into account. ISM is well oiled and tested.
### Todos
- [x] Change relay policy to allow tx.version=2
- [x] fix bip68-sequence-p2p.py tests
- [x] fix bip112-csv-p2p.py tests
### Misc

This softfork _could_ also include SegWit BIP141.

Obviously, needs back-port to `0.12` at least.

btcdrak committed:
Soft fork logic for BIP112

btcdrak committed:
Soft fork logic for BIP113

btcdrak committed:
Soft fork logic for BIP68

btcdrak committed:
RPC test for IsSuperMajority() upgrade blockversion >= 5

NicolasDorier committed:
RPC tests for BIP113 soft fork

NicolasDorier committed:
RPC tests for BIP68 soft fork

NicolasDorier commented:
Tested ACK 1fc8385

maaku commented:
Sticklers for process are going to tell you that you should remove the BIP 113 soft-fork as such transactions are only just now getting marked as non-standard in the 0.12 release. By following this logic the earliest a 113 soft-fork could be applied is after the 0.13 release, at which point in time 0.12 will be the oldest supported version.

btcdrak commented:
@maaku BIP113 mempool policy was deployed since 0.11.2 with the CLTV softfork, so the earliest available softfork is in fact 0.12.1.

btcdrak commented:
@petertodd I structured the commits this way to make it easy to change. So I'm happy to rebase away the specific commits if it's not absolutely _required_.

btcdrak committed:
Policy: allow transaction version 2 relay policy.

This commit introduces a way to gracefully bump the default
transaction version in a two step process.

morcos commented:
a couple comments/questions about the rpc tests:
 - `blockversion5-ism.py` seems extraneous to me.  We already have prior versions of ism tests and in particular the p2p tests added here are testing the ISM logic for version 5 blocks already.
 - For BIP 68 we have the existing `bip68-sequence.py` which tests the logic of the code in a mempool only setting.  I think we need to decide whether that is sufficient when combined with a simple test that the activation logic is working in the bip68 test added here.  But for BIP 112 and 113 it seems we are still missing any tests that exercise the logic, other than the one simple case used to test activation in the rpc tests added here.
 - the RPC tests could have clearer names, for instance the existing bip68 rpc test is also a p2p test, so the distinction in the names is confusing, perhaps `bip68-sequence-activation.py`

morcos commented on qa/rpc-tests/mempool_reorg.py:
```diff
@@ -59,7 +59,8 @@ def run_test(self):
         # Set the time lock
-        timelock_tx = timelock_tx.replace("ffffffff", "11111111", 1)
+        # Disable SequenceLock, Enable nLockTime
+        timelock_tx = timelock_tx.replace("ffffffff", "11111191", 1)
```
in case it saves anyone else time maybe add:
`set most signficant bit (little-endian) to activate LOCKTIME_DISABLE_FLAG`

morcos commented:
quick once over utACK on the code.

although there is a comment in CheckFinalTx about a convention of negative flags meaning consensus enforced rules.  I think that means you need to change the 0 in https://github.com/btcdrak/bitcoin/blob/1fc83850c1308d6f72f357980f66dc9b30f3a4ed/src/main.cpp#L692 to match the result of ISM flags.  But I'm not sure I understand the reasoning for that or if it's necessary.

There does seem to be some missing logic around whether code here or in miner.cpp can actually somehow use STANDARD flags that accidentally don't incorporate consensus flags.  But that's not a new problem to this PR.

btcdrak commented:
I have updated the OP to explain the PR rationale in more detail to address common questions.

morcos commented:
A couple of comments about your PR message:
 - BIP 68 does not depend on BIP 113.  It independently enforces the same semantics, but BIP 113 only applies for lock time locks, not sequence locks.
 - Are you sure about bit masking version numbers for ISM?  It seems to me like that is fraught with problems.

btcdrak commented:
@morcos updated OP.

btcdrak committed:
RPC tests for BIP112 soft fork

NicolasDorier commented:
@morcos @btcdrak CheckFinalTx and removeForReorg always enforce MTP since 0.11.1. To reflect that, I just got rid of the flags https://github.com/NicolasDorier/bitcoin/commit/f1cbaa9a4a3a734601e8c453bcab64a8640b1453

jmcorgan commented:
Initial read through utACK 3bd5c44.  However the additional changes by @NicolasDorier seem unrelated enough to enabling the soft fork logic that I think it would be clearer to have it in a separate PR.

btcdrak commented:
@jmcorgan Yes, we could say the same for the relay policy change for tx version. I personally prefer to keep the commits atomic the it's easy to review while keeping in one PR. The changes are related. Splitting up into a ton of PRs isn't optimal for review vs atomic commits and I believe it's less work in the long run.

maaku commented:
@btcdrak BIP 68 doesn't require or presume BIP 113 -- at least not in the original implementation. It must necessarily use MTP as the start point, but the end point is either nTime or MTP based on activation of BIP 113.

In any case this is a call for @laanwj and @gmaxwell .

morcos commented:
@NicolasDorier I'm sympathetic to trying to remove these flags, but I don't know if should be trying to confuse things by doing it now.   The question is do we want to continue semi-supporting a mempool policy that doesn't enforce MedianTimePast.

@maaku BIP 68 was modified to use MTP semantics for both end points already.  There was discussion about it a couple of months ago and universal agreement.

NicolasDorier commented:
@morcos so what is missing ?

> For BIP 68 we have the existing bip68-sequence.py which tests the logic of the code in a mempool only setting. I think we need to decide whether that is sufficient when combined with a simple test that the activation logic is working in the bip68 test added here. But for BIP 112 and 113 it seems we are still missing any tests that exercise the logic, other than the one simple case used to test activation in the rpc tests added here.

I think this PR should be specific about the softfork logic change, which is tested here.
I propose to move my commit f43e795 + tests about mempool enforcement of BIP112 and 113 in a separate PR, would it be fine for you ?

NicolasDorier commented:
@morcos, I have opened a separate PR for removing the flags in #7574

jtoomim commented on src/main.cpp:
```diff
@@ -2176,6 +2176,14 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
+    // Start enforcing BIP68 (sequence locks) and CHECKSEQUENCEVERIFY (BIP112) for block.nVersion=5
+    // blocks, when 75% of the network has upgraded:
+    int nLockTimeFlags = 0;
+    if (block.nVersion >= 5 && IsSuperMajority(5, pindex->pprev, chainparams.GetConsensus().nMajorityEnforceBlockUpgrade, chainparams.GetConsensus())) {
```
This should be something like

    if ((block.nVersion & 15) >= 5 && IsSuperMajority... )

In order to avoid misinterpreting Version Bits blocks as votes for BIP112. IsSuperMajority would need to be modified slightly as well.

Once this has been done, Bitcoin Classic, XT, and Unlimited can use block versions like 0x300000005 to be able to vote for BIP109 and BIP112 simultaneously.

petertodd replied:
This has been discussed a number of times on the mailing list; what you propose is not a good idea. You can read more from the similar discussion on the mailing list back in Sept in the context of CLTV.

chaosgrid replied:
So what is your proposed fix? Or do you want to risk forking prematurely? You do realize this could put Core on a minority chain in the event that Classic has 50+ % hashing power?

petertodd replied:
We have a proposed fix - versionbits - as well as a backup fix - pseudo-versionbits.

petertodd replied:
We have a proposed fix - versionbits - as well as a backup fix - pseudo-versionbits.

btcdrak commented:
Closing in favour of 7648
